### ISSUE TITLE
Loss of accuracy in weights of large Gauss-Jacobi rules

### ISSUE THREAD

**Issue by pbeckman** [JuliaApproximation/FastGaussQuadrature.jl#130]
While using some large Gauss-Jacobi rules, I found that for highly singular integrands ($\alpha < -0.5$ or so), I'm losing digits in the resulting integrals. As a minimal example, consider
$$\int_{-1}^1 (1-x)^\alpha dx = \frac{2^{\alpha+1}}{\alpha+1}.$$
For a range of $n$'s, we approximate the above integral by summing the weights of an $n$-point Gauss-Jacobi rule.
```julia
using FastGaussQuadrature

a = -0.9
ns = 2 .^ (1:16)
I_true = 2^(a+1)/(a+1)

Is_quad = [sum(gaussjacobi(n, a, 0)[2]) for n in ns]
rel_errs = (I_true .- Is_quad) / I_true
```
which yields the following errors
```julia
n       relative error
----------------------
2       -1.657e-16
4       0.000e+00
8       0.000e+00
16      0.000e+00
32      0.000e+00
64      1.657e-16
128     -5.310e-13
256     7.170e-12
512     5.277e-12
1024    1.242e-11
2048    -1.904e-10
4096    1.459e-09
8192    2.030e-09
16384   -1.294e-08
32768   -4.663e-08
65536   6.886e-08
```
Is this expected behavior? Or perhaps related to [this issue](https://github.com/JuliaApproximation/FastGaussQuadrature.jl/issues/58) posted by @JamesCBremerJr?

Figure 4.5 in Hale and Townsend 2012 (on which I believe this routine is based?) shows 13 to 14 digits of accuracy in the weights for $\alpha = 2, \beta = -0.75$ up to $10^4$ nodes. But if we rerun the above code with $\alpha = -0.75$, we get only 10 digits in the integral for $n = 2^{16}$. 

If we consider larger rules with $\alpha = -0.75$, we get only 8 digits for $n=2^{20}$. And if we take a harsher singularity, e.g. $\alpha = -0.99$, we get only 6 digits for $n=2^{16}$. So it looks like this loss of accuracy eventually (but long before any nodes coincide to machine precision) appears for any $\alpha < -0.5$, and becomes more severe as $\alpha \to -1$. Are such cases known limitations?

**Comment by pbeckman** [comment]
If we take the chebfun weights as ground truth (given that they integrate $f \equiv 1$ to 15 digits) and plot the relative error between the chebfun and FastGaussQuadrature.jl weights, we see that the error is highly concentrated near the endpoints. For example, with $\alpha = -0.99$ and $n = 2^{16},$ we have
![jacobi_error_a99](https://github.com/JuliaApproximation/FastGaussQuadrature.jl/assets/19937363/f8a4bb3c-df32-4c7c-9c20-d2b9314b04e0)
Markers are sized and colored by error size for emphasis. This is a more exaggerated case of exactly what @JamesCBremerJr noted in the [aforementioned issue](https://github.com/JuliaApproximation/FastGaussQuadrature.jl/issues/58).

**Comment by pbeckman** [comment]
The issue is that in the Newton iterations at `gaussjacobi.jl:382`, the recursive formula `innerjacobi_rec` is used instead of the boundary asymptotics, which are implemented as `feval_asy2` in Chebfun. I now have a fork [here](https://github.com/pbeckman/FastGaussQuadrature.jl) with the low order terms in these asymptotics implemented. This gives us back 5 digits or so. See below.
![jacobi_error_a99_new](https://github.com/JuliaApproximation/FastGaussQuadrature.jl/assets/19937363/43eb3ff2-019f-46fc-b8e9-78967d3fd1ca)
However, to get 16 digit accuracy, we need the higher order terms. The higher order coefficients $A_m$ and $B_m$ in equation (3.24) of Hale and Townsend 2012 are computed numerically in Chebfun from recursive integro-differential formulae using Chebyshev interpolation, integration, and differentiation. But these routines are not readily available in FastGaussQuadrature.jl.

Do these Chebyshev routines exist elsewhere in JuliaApproximation? Or (at the risk of bloating this package) should we implement them here?

**Comment by hyrodium** [comment]
Thank you for the detailed comments!
PRs are always welcomed.

>Do these Chebyshev routines exist elsewhere in JuliaApproximation?

I'm not much familiar with Chebyshev polynomials, but maybe [FastChebInterp.jl](https://github.com/JuliaMath/FastChebInterp.jl) helps here?

**Comment by dlfivefifty** [comment]
It should all be doable with ClassicalOrthogonalPolynomials.jl but that depends on this package./.

**Comment by pbeckman** [comment]
Thanks for your responses! 

Unless I'm mistaken, it doesn't appear that Chebyshev integration is currently implemented in ClassicalOrthogonalPolynomials.jl. As an alternative, [ApproxFunOrthogonalPolynomials.jl](https://github.com/JuliaApproximation/ApproxFunOrthogonalPolynomials.jl/tree/master) has both Chebyshev integration and differentiation implemented. However, it is similarly a circular dependency with this package...

In my view, part of the value of FastGaussQuadrature.jl is that it is extremely lightweight, with relatively few dependencies. I can use ApproxFunOrthogonalPolynomials.jl to solve this issue, but it's a heavy (not to mention circular) dependency, so from a software engineering perspective it doesn't seem like a great path forward. It's a hefty price to pay for some basic Chebyshev routines on $N = 10$ points.

The alternative is to use some lookups and reimplement very barebones routines for the necessary computations. This is the path I've taken in [my fork](https://github.com/pbeckman/FastGaussQuadrature.jl). I've saved the 10-point Chebyshev integration and differentiation matrices as constants in `gaussjacobi.jl`, and reimplemented only the necessary barycentric interpolation routine. This is pretty painless and keeps the repository lightweight.

@dlfivefifty as a contributor to all the involved libraries, let me know if you think this is a reasonable solution. If so, I'll try to iron out the remaining numerical issues (all the existing package tests pass, but I still get ~11 digits on worst-case tests related to this issue, not 16) and submit a pull request.

**Comment by pbeckman** [comment]
Ah, I see - I was looking for an `integrate`-like function instead of a `cumsum`.

Your `ChebyshevTransforms.jl` solution makes good sense to me. Given my lack of familiarity with the relevant interfaces in `JuliaApproximation.jl`, I've submitted a pull request in the meantime which resolves this issue and #58. It uses saved integration and differentiation matrices from Chebfun as I mentioned above. It should be easy to substitute more elegant Chebyshev routines once the restructuring of the `JuliaApproximation.jl` package ecosystem that you suggest occurs. But I imagine you may want to engineer that yourself as the manager of those packages.

**Comment by hyrodium** [comment]
Fixed by https://github.com/JuliaApproximation/FastGaussQuadrature.jl/pull/131.
